velox/dwio/common/ColumnVisitors.h

```diff
@@ -514,7 +567,15 @@ ColumnVisitor<T, TFilter, ExtractValues, isDense>::filterFailed() {
 template <typename T, typename TFilter, typename ExtractValues, bool isDense>
 inline void ColumnVisitor<T, TFilter, ExtractValues, isDense>::addResult(
     T value) {
-  values_.addValue(rowIndex_, value);
+  constexpr bool hasHook = !std::is_same_v<
+      typename ColumnVisitor::HookType,
+      velox::dwio::common::NoHook>;
+
+  if (hasHook) {
+    values_.addValue(rowIndex_ + reader_->numScanned(), value);
```

Comment:
I don't think you should adjust here, you can do it in `ExtractToHookWrapper`

Reply:
Good suggestion, I have moved this part of the logic into `ExtractToHookWrapper` as suggested.

Reply:
After removing `ExtractToHookWrapper`, I moved this logic back.

Reply:
Can we set the `rowIndex_` to be already offset by `numScanned`?  That's what I have done in Nimble.  Having different indices would be a source of bug.

Reply:
I tried to modify `rowIndex_` in `ColumnVisitors#process`,:
```
  FOLLY_ALWAYS_INLINE vector_size_t
  process(T value, bool& atEnd, int32_t numScanned = 0) {
    constexpr bool hasHook = !std::is_same_v<
        typename ColumnVisitor::HookType,
        velox::dwio::common::NoHook>;

    if (hasHook) {
      rowIndex_ = rowIndex_ + numScanned;
    }

    if (!TFilter::deterministic) {
      auto previous = currentRow();
      if (velox::common::applyFilter(filter_, value)) {
        filterPassed(value, numScanned);
      } else {
        filterFailed();
      }
      if (++rowIndex_ >= numRows_) {
        atEnd = true;
        return rowAt(numRows_ - 1) - previous;
      }
      return currentRow() - previous - 1;
    }
    // The filter passes or fails and we go to the next row if any.
    if (velox::common::applyFilter(filter_, value)) {
      filterPassed(value, numScanned);
    } else {
      filterFailed();
    }
    if (++rowIndex_ >= numRows_) {
      atEnd = true;
      return 0;
    }
    if (isDense) {
      return 0;
    }
    return currentRow() - rowAt(rowIndex_ - 1) - 1;
  }
```

but the following exception was reported when executing it later:
```
    unknown file: Failure
C++ exception with description "Exception: VeloxRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Retriable: False
Expression: current == testRow
Function: checkAndSkipNulls
File: /data/home/wjameswu/data/code/apache/velox/./velox/dwio/common/ColumnVisitors.h
Line: 198
Stack trace:
# 0  _ZN8facebook5velox7process10StackTraceC1Ei
# 1  _ZN8facebook5velox14VeloxException5State4makeIZNS1_C4EPKcmS5_St17basic_string_viewIcSt11char_traitsIcEES9_S9_S9_bNS1_4TypeES9_EUlRT_E_EESt10shared_ptrIKS2_ESA_SB_
# 2  _ZN8facebook5velox14VeloxExceptionC1EPKcmS3_St17basic_string_viewIcSt11char_traitsIcEES7_S7_S7_bNS1_4TypeES7_
# 3  _ZN8facebook5velox17VeloxRuntimeErrorC2EPKcmS3_St17basic_string_viewIcSt11char_traitsIcEES7_S7_S7_bS7_
# 4  _ZN8facebook5velox6detail14veloxCheckFailINS0_17VeloxRuntimeErrorENS1_22CompileTimeEmptyStringEEEvRKNS1_18VeloxCheckFailArgsET0_
# 5  _ZN8facebook5velox7parquet13StringDecoder15readWithVisitorILb1ENS0_4dwio6common13ColumnVisitorIN5folly5RangeIPKcEENS0_6common10AlwaysTrueENS5_20ExtractToGenericHookELb0EEEEEvPKmT0_i
# 6  _ZN8facebook5velox7parquet10PageReader11callDecoderINS0_4dwio6common13ColumnVisitorIN5folly5RangeIPKcEENS0_6common10AlwaysTrueENS5_20ExtractToGenericHookELb0EEELi0EEEvPKmRbT_i
# 7  _ZN8facebook5velox7parquet10PageReader15readWithVisitorINS0_4dwio6common13ColumnVisitorIN5folly5RangeIPKcEENS0_6common10AlwaysTrueENS5_20ExtractToGenericHookELb0EEEEEvRT_
# 8  _ZN8facebook5velox7parquet11ParquetData15readWithVisitorINS0_4dwio6common13ColumnVisitorIN5folly5RangeIPKcEENS0_6common10AlwaysTrueENS5_20ExtractToGenericHookELb0EEEEEvT_
# 9  _ZZN8facebook5velox7parquet18StringColumnReader4readEiN5folly5RangeIPKiEEPKmENKUlT_E0_clINS0_4dwio6common13ColumnVisitorINS4_IPKcEENS0_6common10AlwaysTrueENSE_20ExtractToGenericHookELb0EEEEEDaSA_
# 10 _ZN8facebook5velox4dwio6common33StringColumnReadWithVisitorHelperILb1ELb0EE10readHelperINS0_6common10AlwaysTrueELb0ENS2_20ExtractToGenericHookEZNS0_7parquet18StringColumnReader4readEiN5folly5RangeIPKiEEPKmEUlT_E0_EEvPNS6_6FilterET1_T2_
# 11 _ZN8facebook5velox4dwio6common33StringColumnReadWithVisitorHelperILb1ELb0EEclIZNS0_7parquet18StringColumnReader4readEiN5folly5RangeIPKiEEPKmEUlT_E0_EEDaOSF_
# 12 _ZN8facebook5velox7parquet18StringColumnReader4readEiN5folly5RangeIPKiEEPKm
# 13 _ZN8facebook5velox4dwio6common12ColumnLoader12loadInternalEN5folly5RangeIPKiEEPNS0_9ValueHookEiPSt10shared_ptrINS0_10BaseVectorEE
# 14 _ZN8facebook5velox12VectorLoader4loadEN5folly5RangeIPKiEEPNS0_9ValueHookEiPSt10shared_ptrINS0_10BaseVectorEE
# 15 _ZNK8facebook5velox10LazyVector4loadEN5folly5RangeIPKiEEPNS0_9ValueHookE
# 16 _ZN8facebook5velox4dwio6common17E2EFilterTestBase17checkLoadWithHookILNS0_8TypeKindE7EEEbRKSt6vectorISt10shared_ptrINS0_9RowVectorEESaIS9_EEPS8_iS7_INS0_10BaseVectorEERKS6_ImSaImEEi
# 17 _ZZN8facebook5velox4dwio6common17E2EFilterTestBase12loadWithHookERKSt6vectorISt10shared_ptrINS0_9RowVectorEESaIS7_EEPS6_iS5_INS0_10BaseVectorEERKS4_ImSaImEEiENKUlvE_clEv
# 18 _ZN8facebook5velox4dwio6common17E2EFilterTestBase12loadWithHookERKSt6vectorISt10shared_ptrINS0_9RowVectorEESaIS7_EEPS6_iS5_INS0_10BaseVectorEERKS4_ImSaImEEi
# 19 _ZN8facebook5velox4dwio6common17E2EFilterTestBase14readWithFilterESt10shared_ptrINS0_6common8ScanSpecEERKNS2_12MutationSpecERKSt6vectorIS4_INS0_9RowVectorEESaISD_EERKSB_ImSaImEERmbb
# 20 _ZN8facebook5velox4dwio6common17E2EFilterTestBase22testReadWithFilterLazyERKSt10shared_ptrINS0_6common8ScanSpecEERKNS2_12MutationSpecERKSt6vectorIS4_INS0_9RowVectorEESaISF_EERKSD_ImSaImEE
# 21 _ZN8facebook5velox4dwio6common17E2EFilterTestBase15testFilterSpecsERKSt6vectorISt10shared_ptrINS0_9RowVectorEESaIS7_EERKS4_INS2_10FilterSpecESaISC_EE
# 22 _ZN8facebook5velox4dwio6common17E2EFilterTestBase18testNoRowGroupSkipERKSt6vectorISt10shared_ptrINS0_9RowVectorEESaIS7_EERKS4_INSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaISH_EEi
# 23 _ZN8facebook5velox4dwio6common17E2EFilterTestBase12testScenarioERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt8functionIFvvEEbRKSt6vectorIS9_SaIS9_EEib
# 24 _ZN13E2EFilterTest13testWithTypesERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt8functionIFvvEEbRKSt6vectorIS5_SaIS5_EEi
# 25 _ZN31E2EFilterTest_stringDirect_Test8TestBodyEv
# 26 _ZN7testing8internal38HandleSehExceptionsInMethodIfSupportedINS_4TestEvEET0_PT_MS4_FS3_vEPKc
# 27 _ZN7testing8internal35HandleExceptionsInMethodIfSupportedINS_4TestEvEET0_PT_MS4_FS3_vEPKc
# 28 _ZN7testing4Test3RunEv
# 29 _ZN7testing8TestInfo3RunEv
# 30 _ZN7testing9TestSuite3RunEv
# 31 _ZN7testing8internal12UnitTestImpl11RunAllTestsEv
# 32 _ZN7testing8internal38HandleSehExceptionsInMethodIfSupportedINS0_12UnitTestImplEbEET0_PT_MS4_FS3_vEPKc
# 33 _ZN7testing8internal35HandleExceptionsInMethodIfSupportedINS0_12UnitTestImplEbEET0_PT_MS4_FS3_vEPKc
# 34 _ZN7testing8UnitTest3RunEv
# 35 _Z13RUN_ALL_TESTSv
# 36 main
# 37 __libc_start_main
# 38 _start
" thrown in the test body.
```

Reply:
You should not have different treatment for hook and non-hook case for the indices.  They should receive the same and column visitor just apply the same indices.